op-service/client/retry.go

```diff
@@ -53,12 +54,68 @@ func (b *retryingClient) CallContext(ctx context.Context, result any, method str
 	})
 }
 
-func (b *retryingClient) BatchCallContext(ctx context.Context, batch []rpc.BatchElem) error {
+// pendingReq combines BatchElem information with the index of this request in the original []rpc.BatchElem
+type pendingReq struct {
+	// req is a copy of the BatchElem individual request to make.
+	// It never has Result or Error set as it gets copied again as part of being passed to the underlying client.
+	req rpc.BatchElem
+
+	// idx tracks the index of the original BatchElem in the supplied input array
+	// This can then be used to set the result on the original input
+	idx int
+}
+
+func (b *retryingClient) BatchCallContext(ctx context.Context, input []rpc.BatchElem) error {
+	// Add all BatchElem to the initial pending set
+	// Each time we retry, we'll remove successful BatchElem for this list so we only retry ones that fail.
+	pending := make([]*pendingReq, len(input))
+	for i, req := range input {
+		pending[i] = &pendingReq{
+			req: req,
+			idx: i,
+		}
+	}
 	return backoff.DoCtx(ctx, b.retryAttempts, b.strategy, func() error {
 		cCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
 		defer cancel()
+
+		batch := make([]rpc.BatchElem, len(pending))
+		for i, req := range pending {
+			batch[i] = req.req
+		}
 		err := b.c.BatchCallContext(cCtx, batch)
-		return err
+		if err != nil {
+			// Whole call failed, retry all pending elems again
+			return err
+		}
+		var failed []*pendingReq
+		var combinedErr error
+		for i, elem := range batch {
+			req := pending[i]
+			idx := req.idx // Index into input of the original BatchElem
+
+			// Set the result on the original batch to pass back to the caller in case we stop retrying
+			input[idx].Error = elem.Error
+			input[idx].Result = elem.Result
+
+			// If the individual request failed, add it to the list to retry
+			if elem.Error != nil {
+				// Need to retry this request
+				failed = append(failed, req)
+				if combinedErr == nil {
+					combinedErr = elem.Error
+				} else {
+					// Not using fmt.Errorf because lint rules disallow using %v for errors and you can't have two %w
+					// Replace with errors.Join when we move to go 1.20
+					combinedErr = errors.New(combinedErr.Error() + "\n" + elem.Error.Error())
+				}
+			}
+		}
+		if len(failed) > 0 {
+			pending = failed
+			return combinedErr
+		}
+		return nil
```

Comment:
I think we can use [multierror](https://pkg.go.dev/github.com/hashicorp/go-multierror#section-readme) here to append errors nicely. I believe it's a fairly standard golang package also used in geth (if my memory's not failing me).
```suggestion
		var combinedErr error
		for i, elem := range batch {
			req := pending[i]
			idx := req.idx // Index into input of the original BatchElem

			// Set the result on the original batch to pass back to the caller in case we stop retrying
			input[idx].Error = elem.Error
			input[idx].Result = elem.Result

			// If the individual request failed, add it to the list to retry
			if elem.Error != nil {
				// Need to retry this request
				failed = append(failed, req)
				combinedErr = multierror.Append(combinedErr, elem.Error)
			}
		}
		if len(failed) > 0 {
			pending = failed
			return combinedErr
		}
		return nil
```

Reply:
IDK if geth uses multierror, but we used hashicorp's version before & there is now a version in the standard library.

Reply:
Standard library has `errors.Join` but only in go 1.20 which we don't use yet.
More recent go also lets you use two `%w` in `fmt.Errorf` but the version of go we use doesn't. (I had some arguments with CI about these things....)

We do use hashicorp multierror already so have switched to that which is a nice cleanup.

Reply:
maybe time to get onto go 1.20. 1.19 will EOL this fall